### ISSUE TITLE
Pitstop II not detecting PCjr

### ISSUE THREAD

**Issue by rderooy** [dosbox-staging/dosbox-staging#2266]
### Are you using the latest Dosbox-Staging Version?

- [X] I have checked releases and am using the latest release.

### Different version than latest?

git version

### What Operating System are you using?

Linux x86_64

### Have you checked that no other similar issue already exists?

- [X] I have searched and not found similar issues.

### A clear and concise description of what the bug is.

Running the early PC booter game "Pitstop II" by Epyx with `machine=pcjr` displays CGA graphics instead of PCjr graphics.

It appears the game is not able to detect a PCjr and defaults to CGA. But when run on 86Box or MESS, it is capable of displaying PCjr graphics.

https://nerdlypleasures.blogspot.com/2023/01/cga-and-intended-incidental-and-just.html

When starting the game, it asks to press 1 or 2. According to the games manual, 1 is composite and 2 is RGB output.

### Steps to reproduce the behaviour.

Explain how to reproduce
1. 
2.
3.

### Code of Conduct & Contributing Guidelines

- [X] Yes, I agree.

**Comment by kcgen** [comment]
Thanks @rderooy 

Some comments in the article:

> "it was discovered only recently that Pitstop II will not display the PCjr.'s intended RGB graphics as a real PCjr. does unless the emulator supports the PCjr. BIOS"

> "DOSBox does not support booting from the PCjr. BIOS, 86Box and MESS do."

Perhaps the BIOS sets up some kind of palette handler and maybe sets some important graphics bits (like are emulated in `src/ints/bios.cpp`).

**Comment by rderooy** [comment]
> Perhaps the BIOS sets up some kind of palette handler and maybe sets some important graphics bits (like are emulated in `src/ints/bios.cpp`).

If that were the case, I'm guessing other games would also be effected and it would be a known issue. Instead I suspect it is simply looking for some kind of signature in the BIOS, failing to find it, and assuming CGA.

**Comment by NRS-NewRisingSun** [comment]
https://www.vogons.org/viewtopic.php?f=41&t=50417&p=534365

No. 1

**Comment by Torinde** [comment]
joncampbell123/dosbox-x/issues/2971

**Comment by johnnovak** [comment]
Just play the C64 version 😎 Sure, it would be nice to be able to run it if it's not too hard, but not exactly a great loss...

**Comment by rderooy** [comment]
@NRS-NewRisingSun I had a moment to compare the "Method 1" mentioned in the Vogons post with the code in dosbox-staging.

The only difference the code posted on Vogons seems to make, is to use a different BIOS_DEFAULT_RESET_LOCATION for regular PC and PCjr.

But if you look at the dosbox-staging `include/bios.h` file, it already adjusts the location based on machine type:
```cpp
include/bios.h:#define BIOS_DEFAULT_RESET_LOCATION		(RealMake(0xf000,(machine==MCH_PCJR)?0x0043:0xe05b))
```

**Comment by NRS-NewRisingSun** [comment]
Indeed. The game _already_ properly detects the PCjr in DOSBox-Staging. OP is merely mistaken about what that PCjr support will look like. Even if the game detects PCjr, it will ask you about Composite vs. EGA, and it will only produce PCjr-specific graphics if you press "1" for Composite. This make no sense, but that's just how the game is programmed. The PCjr-specific graphics will then look like this:

![PitstopII-PCjr](https://github.com/dosbox-staging/dosbox-staging/assets/32988436/871a28d3-8e34-4bed-9c10-977c40a6e1dd)

**Comment by rderooy** [comment]
Thanks @NRS-NewRisingSun 

This is what I get, with-and-without the int10 patch.

The game was started with either:
```
dosbox --noprimaryconf --nolocalconf --machine pcjr -c 'composite=off' -c 'boot PITSTOP2.IMG'
```
or
```
dosbox --noprimaryconf --nolocalconf --machine pcjr -c 'composite=on' -c 'boot PITSTOP2.IMG'
```

|          | Option 1 | Option 2 |
|--------|--------|--------|
| composite=off | ![pitstop2-pcjr-option1-compositeOff](https://github.com/dosbox-staging/dosbox-staging/assets/17701098/b8eb0191-eb4b-49ad-94ec-7578149e37af)  | ![pitstop2-pcjr-option2-compositeOff](https://github.com/dosbox-staging/dosbox-staging/assets/17701098/9b89fe1a-ad3b-473f-ae86-6237396a93e8) |
| composite=on | ![pitstop2-pcjr-option1-compositeOn](https://github.com/dosbox-staging/dosbox-staging/assets/17701098/919c321d-66a3-4bd8-9667-b0b87833b96f) | ![pitstop2-pcjr-option2-compositeOn](https://github.com/dosbox-staging/dosbox-staging/assets/17701098/f3a9872c-56af-4234-97c7-88e339d59e5f) |
| composite=off, int10 patch | ![pitstop2-pcjr-int10patch-option1-compositeOff](https://github.com/dosbox-staging/dosbox-staging/assets/17701098/ef675d40-03c9-4674-a717-bb3a2fb91351) | ![pitstop2-pcjr-int10patch-option2-compositeOff](https://github.com/dosbox-staging/dosbox-staging/assets/17701098/eb07dc14-1b6f-43d9-a45c-a8cef12f5c7c) |
| composite=on, int10 patch | ![pitstop2-pcjr-int10patch-option1-compositeOn](https://github.com/dosbox-staging/dosbox-staging/assets/17701098/cb4c9da7-5c74-4f0d-b9f0-6f1aafbe94dc) | ![pitstop2-pcjr-int10patch-option2-compositeOn](https://github.com/dosbox-staging/dosbox-staging/assets/17701098/a32c07d4-fda3-4c42-bb85-e9364e0247bc) |

According to the above linked blog post, Option 1 is for composite displays, and Option 2 is for RGB. But I find it strange that the output of Option 1 never changes, regardless of composite setting.

**Comment by rderooy** [comment]
And indeed this Spanish manual claims that Option 1 is "video composito", and Option 2 is "RGB". https://archive.org/details/pitstop-ii/page/n5/mode/2up

**Comment by NRS-NewRisingSun** [comment]
"But I find it strange that the output of Option 1 never changes, regardless of composite setting." That is because the game selects 160x200 mode via BIOS (Int 10h AH=00 AL=88h) in that situation, and DOSBox-Staging's composite emulation does not affect 160x200 mode just as it does not affect text mode, only 320x200 and 640x200 modes. There is a logic to this, as 160x200 mode normally does not need composite emulation since there are no desirable artifact colors involved, only an uglier palette. But I agree that it seems confusing. Reengine's rewritten composite emulation can be applied to any mode, but it never was updated for PCjr artifacting (unless I missed it).

The main issue here is now solved -- PCjr was always correctly detected. Two follow-up issues persist:
1. Whether to include the Int 10h AH=0Bh patch that I mentioned or not. Yes, because it's what the PCjr BIOS does, no, because it's a bug, not a feature, of the PCjr BIOS.
2. Whether composite emulation should be modified to be applicable to 160x200 mode as well.

**Comment by rderooy** [comment]
Thanks for the explanation @NRS-NewRisingSun 

One thing to note about text modes and composite, is that while it is disabled by default, even with `composite=on`, you can force it on by cycling through the composite options with Ctrl-F12 (although any screen changes will reset it again). But this does not work with the 160x200 mode.

**Comment by johnnovak** [comment]
Well, at least now I know of a single game that uses the 160x200 mode 😄 That comes in handy for regression testing. Tips on any other games that use this mode? Early Sierra AGI games actually use 320x200.

**Comment by johnnovak** [comment]
While we're at this, can you guys explain to me what the difference between `composite = on` and `auto` is? What does `auto` even mean—you either had a composite monitor, or you didn't no?

**Comment by NRS-NewRisingSun** [comment]
composite=auto means 640x200 with color burst enabled ($3D8=$1A) is always displayed as composite and nothing else, in particular not 320x200 mode.

"Well, at least now I know of a single game that uses the 160x200 mode"  There's a ton of them, actually. The non-enhanced versions of Maniac Mansion and Zak McKracken, Oil's Well, Demon Attack, Touchdown Football, Bruce Lee, Boulder Dash (II), Pitfall II, River Raid, Ghostbusters... Ghostbusters is the most interesting one because it uses a raster interrupt to change between 160x200x16 and 320x200x4 mode mid-screen, which works in DOSBox-Staging with RGB display but fails horribly with composite emulation on.

**Comment by johnnovak** [comment]
Thanks @NRS-NewRisingSun, very informative!
